lib/bolt/cli.rb

```diff
@@ -379,710 +309,563 @@ def validate(options)
+                    else
+                      Bolt::Project.find_boltdir(Dir.pwd)
+                    end
 
-    def query_puppetdb_nodes(query)
-      puppetdb_client.query_certnames(query)
-    end
+          config = Bolt::Config.from_project(project, options)
+
+          # TODO: Should this be an instance variable? Outputter is specific
+          #       to the CLI, so might make sense to have it be a property of
+          #       the class.
+          @outputter = Bolt::Outputter.for_format(
+            config.format,
+            config.color,
+            options[:verbose],
+            config.trace,
+            config.spinner
+          )
+
+          # TODO: Should this be an instance variable? Rerun is specific
+          #       to the CLI, so might make sense to have it be a property of
+          #       the class.
+          @rerun = Bolt::Rerun.new(config.rerunfile, config.save_rerun)
+
+          # TODO: Should this be an input to the executor? Might make sense
+          #       to subscribe it to the executor, as it's an implementation
+          #       of the CLI tool and might not make sense for the API.
+          analytics = begin
+            client = Bolt::Analytics.build_client(config.analytics)
+            client.bundled_content = bundled_content(options)
+            client
+          end
 
-    def warn_inventory_overrides_cli(opts)
-      inventory_source = if ENV[Bolt::Inventory::ENVIRONMENT_VAR]
-                           Bolt::Inventory::ENVIRONMENT_VAR
-                         elsif config.inventoryfile
-                           config.inventoryfile
-                         elsif File.exist?(config.default_inventoryfile)
-                           config.default_inventoryfile
-                         end
+          # TODO: Configure logger with a separate method call?
+          Bolt::Logger.configure(config.log, config.color, config.disable_warnings)
+          Bolt::Logger.stream = config.stream
+          Bolt::Logger.analytics = analytics
+          Bolt::Logger.flush_queue
+
+          executor = Bolt::Executor.new(
+            config.concurrency,
+            analytics,
+            options[:noop],
+            config.modified_concurrency,
+            config.future
+          )
+
+          pal = Bolt::PAL.new(
+            Bolt::Config::Modulepath.new(config.modulepath),
+            config.hiera_config,
+            config.project.resource_types,
+            config.compile_concurrency,
+            config.trusted_external,
+            config.apply_settings,
+            config.project
+          )
+
+          plugins = Bolt::Plugin.setup(config, pal, analytics)
+
+          inventory = Bolt::Inventory.from_config(config, plugins)
+
+          log_outputter = Bolt::Outputter::Logger.new(options[:verbose], config.trace)
+
+          #
+          # FINALIZING SETUP
+          #
+
+          check_gem_install
+
+          # TODO: Discuss possibility of Bolt 4.0 having CLI overrides take
```

Comment:
I think the big issue here is target-specific config vs. "global" config, but since we resolved the inventory all at once we don't know what came from target-specific config or top-level config, so everything has to override the CLI. Now that top-level inventory config is supported in the defaults directory which does get overridden by the CLI, documenting that inventory config can go there and then be overridden seems good + then we can point to those docs in the message. But I don't think having "top-level" config from the CLI override specific options makes sense.